### PR TITLE
Support init from numpy with any number of dimensions

### LINKED ISSUE #3224: Support init from NumPy arrays with more than 4 dimensions?
I ran into this today, is there any reason (apart from code size and compile time) against supporting, e.g., also 5 and 6 dims?

https://github.com/scipp/scipp/blob/1c065b166ec06464ad166859a05937e6b18e101c/lib/python/numpy.h#L170-L184

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #3224

# Changes

- Supports any number of dimensions without stamping out more templates.
- Parallelised for any number of dimensions.

`py::array` unfortunately has no interface indexing with a run-time-determined number of indices. So I went through the underlying buffer. 

# Performance

I did some benchmarks. For low dimensions (1, 2) the new code performs worse than the old code. E.g., using
```python
from timeit import timeit
import scipp as sc
import numpy as np

times = {}
for ndim in range(1, 4):
    a = np.zeros([1000]*ndim)
    dims = [f'dim{d}' for d in range(ndim)]
    times[ndim] = timeit(lambda: sc.array(dims=dims, values=a), number=10)

print(times)
```
I get **old**:
```python
{1: 0.00020559499898809008, 2: 0.006096607001381926, 3: 11.701265682000667}
```
**new**:
```python
{1: 0.0010022530004789587, 2: 0.004268341999704717, 3: 4.414793544001441}
```
This is quite a significant slowdown.

But for ndim=3,4 I found speedups.

Is this bad enough to warrant keeping the special cases for low dimensions?

### CODE REVIEW

**Comment by jl-wynen** [comment]
The test failure only happens with parallelisation. And only with non-c-contiguous buffers. I don't know yet what is wrong.

**Comment by SimonHeybrock** [comment]
> I get **old**:
> 
> ```python
> {1: 0.00020559499898809008, 2: 0.006096607001381926, 3: 11.701265682000667}
> ```
> 
> **new**:
> 
> ```python
> {1: 0.0010022530004789587, 2: 0.004268341999704717, 3: 4.414793544001441}
> ```
> 
> This is quite a significant slowdown.
> 
> But for ndim=3,4 I found speedups.
> 
> Is this bad enough to warrant keeping the special cases for low dimensions?

- Can you clarify if this is proportional to the number of elements, or a constant overhead?
- In either case (but certainly in the first), I think it does warrant keeping the special cases.

**Comment by jl-wynen** [comment]
I did a more thorough benchmark using
```python
import json
from itertools import product
from rich import print
from timeit import repeat
import scipp as sc
import numpy as np
import math
import sys

volume = lambda s: math.prod(map(int, s))

def shapes(ndim):
    yield from filter(lambda s: volume(s) < 10**8, product(*(np.logspace(0, 6, 6).astype(int)[::-1] for _ in range(ndim))))

number = 30
times = {}

for ndim in range(1, 5):
    print(f'{ndim=}')
    dims = [f'dim{d}' for d in range(ndim)]

    times[ndim] = {}
    for shape in shapes(ndim):
        a = np.random.random(shape)
        vol = volume(shape)
        if vol in times[ndim]:
            continue
        times[ndim][vol] = min(repeat(lambda: sc.array(dims=dims, values=a), number=number, repeat=5)) / number

    times[ndim] = {vol: times[ndim][vol] for vol in sorted(times[ndim].keys())}

with open(f'{sys.argv[1]}.json', 'w') as f:
    json.dump(times, f)
```
which gives
benchmark


The x-axes show the total volume of the created variable. There are kinks and jumps because there are many different spaes involved.

Overall, there is a significant pessimisation for small arrays but a gain for large ones. Interestingly, for ndim=2, the two approaches are essentially the same. This is because this case was already parallelised in the old implementation. With a serial build, the new implementation produces these times:
benchmark-ser
Still not as good as the old one but a lot closer.

What do you think?

**Comment by SimonHeybrock** [comment]
> What do you think?

I think log-log scale is evil and hides the true difference. Can you make the Y-axis linear, and maybe plot the time per element, or the memory bandwidth (sum of reads+writes in GByte/second).

File: lib/python/numpy.h [hunk 1]
```diff
@@ -85,54 +76,96 @@ auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
...
+
+template <bool convert, class T, class Dst>
+void copy_flattened(const typed_buffer<T> &src, Dst &dst) {
+  if (src.ndim == 0)
+    copy_flattened_0d<convert>(src, dst);
+  else {
+    const auto src_stride = src.stride(0);
+    const auto dst_stride = inner_volume(src);
+    core::parallel::parallel_for(
+        core::parallel::blocked_range(0, src.shape[0]), [&](const auto &range) {
```
Comment (SimonHeybrock) [inline lib/python/numpy.h:162]: Can you try setting a better grainsize here, see comment in https://github.com/scipp/scipp/blob/351eefd7c09b03991c22e175477929767caac266/lib/core/include/scipp/core/parallel-tbb.h#L16-L27

File: tests/variable_creation_test.py [hunk 2]
```diff
@@ -321,6 +321,78 @@ def test_array_empty_dims():
...
+    if dim == 0:
+        return array[::2]
+    elif dim == 1:
+        return array[:, ::2]
+    elif dim == 2:
+        return array[:, :, ::2]
+    elif dim == 3:
+        return array[:, :, :, ::2]
+    elif dim == 4:
+        return array[:, :, :, :, ::2]
```
Comment (SimonHeybrock) [inline tests/variable_creation_test.py:358]: Can you add a test with multiple such slices, as well as negative strides?

File: lib/python/numpy.h (hunk 1 above)
Comment (SimonHeybrock) [inline lib/python/numpy.h:162]: It seems this approach will be very suboptimal if we either have `shape[0]==1` (or very small), or `shape[-1]==1` (or very small). Can you benchmark those two cases as well? In particular the 2-D cases, i.e., (N, 1) and (1, N). The old implementation likely had a similar problem. I wonder if one could `squeeze` such dims?

Or maybe we can move the `parallel_for` call into `copy_flattened_middle_dims`, and make it conditional based on the current dim's size, and whether any out dim was processed in parallel already?

File: docs/about/release-notes.rst [hunk 3]
```diff
@@ -35,6 +35,8 @@ Features
 
 * Operations on binned variables with a data array content buffer now also work with slices of binned variables `#3282 <https://github.com/scipp/scipp/pull/3282>`_.
 * Reduced per-bin overhead in many operations involving binned variables by about 20 milliseconds per 1 million bins `#3282 <https://github.com/scipp/scipp/pull/3282>`_.
+* It is now possible to construct Scipp variables from Numpy arrays with up to 6 dimensions for arbitrary memory layouts and any number of dimensions for c-contiguous memory layouts.
+  The limit used to be ``ndim <= 4`` `#3284 <https://github.com/scipp/scipp/pull/3284>`_.
```
Comment (SimonHeybrock) [inline docs/about/release-notes.rst:39]: Mention improved performance?

File: lib/python/numpy.h [hunk 4]
```diff
@@ -157,36 +238,60 @@ bool memory_overlaps(const py::array_t<T> &data, const View &view) {
...
     case 4:
-      return copy_flattened_4d<convert>(src_, std::forward<View>(dst_));
+      return copy_flattened_4d<convert>(src_, dst);
+    case 5:
+      return copy_flattened_5d<convert>(src_, dst);
+    case 6:
+      return copy_flattened_6d<convert>(src_, dst);
     default:
       throw std::runtime_error("Numpy array has more dimensions than supported "
                                "in the current implementation.");
```
Comment (SimonHeybrock) [inline lib/python/numpy.h:288]: Do you want to suggest in the error message to make a copy to a c-contiguous array as a workaround?

File: lib/python/numpy.h [hunk 5]
```diff
@@ -85,54 +76,144 @@ auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
...
+      });
+}
+
+template <bool convert, class T, class Dst>
+void copy_flattened(const py::array_t<T> &src_array, Dst &dst) {
+  const auto src_buffer = src_array.request();
+  auto src = reinterpret_cast<const T *>(src_buffer.ptr);
+  const auto begin = dst.begin();
+  core::parallel::parallel_for(
+      core::parallel::blocked_range(0, src_buffer.size, 10000),
```
Comment (SimonHeybrock) [inline lib/python/numpy.h:210]: The 10000 shows up multiple times. Can you give it a name?

File: lib/python/numpy.h [hunk 6]
```diff
@@ -85,54 +76,144 @@ auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
...
+            for (scipp::index k = 0; k < src.shape(2); ++k)
+              for (scipp::index l = 0; l < src.shape(3); ++l)
+                for (scipp::index m = 0; m < src.shape(4); ++m)
+                  for (scipp::index n = 0; n < src.shape(5); ++n, ++it)
+                    copy_element<convert>(src(i, j, k, l, m, n), *it);
+      });
+}
+
+template <bool convert, class T, class Dst>
+void copy_flattened(const py::array_t<T> &src_array, Dst &dst) {
```
Comment (SimonHeybrock) [inline lib/python/numpy.h:205]: What is the difference between this and the `1d` function? Can one use the other?

File: lib/python/numpy.h [hunk 7]
```diff
@@ -85,54 +76,144 @@ auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
...
+        }
+      });
+}
+
+template <bool convert, class T, class Dst>
+void copy_flattened_2d(const py::array_t<T> &src_array, Dst &dst) {
+  const auto src = src_array.template unchecked<2>();
+  const auto begin = dst.begin();
+  core::parallel::parallel_for(
+      core::parallel::blocked_range(0, src.shape(0), 10000),
```
Comment (SimonHeybrock) [inline lib/python/numpy.h:126]: Isn't this disabling multi-threading if `shape[0] < 10000`? Shouldn't we use the default setup (which, iirc, is `shape[0] / 24`, i.e., threaded for `shape[0]>1`? Or is this harmful in more important cases?

Same comment applies to 3d+ cases.

File: lib/python/numpy.h (hunk 5 above)
Comment (jl-wynen) [reply lib/python/numpy.h:210]: Can do. But the optimum would probably be different for each function.

File: lib/python/numpy.h (hunk 6 above)
Comment (jl-wynen) [reply lib/python/numpy.h:205]: The 1d function works for any strides by indexing the array directly. The flattened function 'indexes' a pointer and only works for c-contiguous data.

**Review by SimonHeybrock — APPROVED** [review]
